cloud_storage: stop shared reader on exception

VladLazar opened the pull request:
### Cover letter
Previously, when an exception was thrown on the read path,
partition_record_batch_reader_impl would simply reassing the reader
without stopping. This violates the shared reader's invariant that it
must be stopped before the destructor runs.

This patch corrects the mistake by also stopping the reader when handling
an unknown exception.

I tested this by inserting a throw on the read path and checking that the
assertion which checks for the reader being stopped does not fire.

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [ ] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [X] v22.2.x
- [X] v22.1.x
- [X] v21.11.x

<!-- don't ship user breaking changes. Ping PMs for help with user visible changes  -->

### Release notes
* none

VladLazar commented:
Urrgh. I changed to handling bad_allocs only. Catching the generic std::exception led to stopping the reader twice.

dotnwat commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -190,16 +191,19 @@ class partition_record_batch_reader_impl final
+        }
 
-            throw;
+        if (bad_alloc_exception_ptr) {
+            co_await set_end_of_stream();
+            std::rethrow_exception(bad_alloc_exception_ptr);
```
i don't know enough about what is going on here to really judge, but i'm having trouble understanding what is happening. the title says we are stopping the reader on exception, but it looks like we are only stopping it on bad_alloc. on the other hand, in the previous version, any exception appeared to reset the reader (but not stop it), and now that isn't done for any exception.

maybe a few examples or a link to an issue with more detail would be helpful. i'm also open to the idea that i have no idea what i'm talking about and shouldn't comment :)

VladLazar replied:
You're right. The code and commit message got out sync. I've updated both now and hopefully it makes sense.

jcsp commented:
I can believe that the previous change had an issue with destroying _reader without stopping it (approving PR on that basis), but have to confess I'm not 100% on what the general behavior is here:
- do we stop the reader nicely somewhere else before the `_reader = {}` on the gate_closed_exception path?
- we could still catch exceptions generically (not just bad_alloc) if the reader had a flag or something to remember whether it had already been stopped, right?

Neither of the above has to block this, just advertising the gaps in my knowledge.

ztlpn commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -190,16 +191,19 @@ class partition_record_batch_reader_impl final
+              "bad_alloc thrown while reading from remote_partition: {}",
               e.what());
-            _it = _end;
-            _reader = {};
```
I wonder if calling `remote_partition::evict_reader` instead of simply resetting it would solve the problem

ztlpn replied:
Although synchronously closing it with `partition_record_batch_reader_impl::set_end_of_stream` looks okay too and more straightforward.

VladLazar replied:
I'm not sure it would. If the reader is returned and handed out before the eviction takes place, the original issue can still happen. Pretty unlikely though.

ztlpn reviewed (commented):
+1 to the question on why we only handle bad_allocs and not other exceptions

ztlpn commented:
> do we stop the reader nicely somewhere else before the _reader = {} on the gate_closed_exception path?

@jcsp Looks suspicious indeed. My first question when reading this is what gate this refers to (too many gates lol).

It can't refer to the partition gate because it is held for the entire duration of the partition reader lifetime.

Maybe it refers to the `remote_segment_batch_reader` gate? But where is the concurrency if the segment reader always has a single owner and is read sequentially and then stopped? Can't remember, maybe this gate is not needed at all...

Or maybe it refers to the remote segment gate. That would mean that someone dehydrated the segment under our feet. I wonder if we could test if this scenario is possible and if it results in the same assert as we are seeing now.

Anyway, if would be great if someone figures this out and leaves a comment explaining it. 😅

VladLazar commented:
Answering a few questions from the comments:
1. Which gate does the gate_closed_exception refer to and why is it needed?
It's the `remote_segment_batch_reader` [gate](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote_segment.h#L262).

As Alexey pointed out, I also don't think there is any concurrency on the `remote_segment_batch_reader`. Readers may be re-used, but not concurrently. Once the `remote_partition` hands out a reader, it will [delete it](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote_partition.cc#L740), so it won't be used concurrently.

The gate exists in order to allow the existing reads and writes to complete, before stopping the `remote_segment_batch_reader`. There's a few scenarios where the reader could be stopped from higher up in the cloud_storage read chain. One example is an [abort being requested](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote_partition.cc#L83). That will lead to the leader being stopped and the gate ensures that the current reads/writes are still served.

2. Do we stop the reader nicely somewhere else before the _reader = {} on the gate_closed_exception path?
The gate in question can only be closed from the [reader's stop method](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote_segment.cc#L980) and that also includes stopping the reader. This means that, on the next read attempt when a gate_closed_exception is thrown we can assume the reader is stopped.

3. Why do we only handle bad_allocs and not other exceptions?
Initially, I handled all exceptions (caught via std::exception), but that meant that I was also catching the exception thrown [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote_partition.cc#L177) in the same function. That exception will only be thrown after the reader is already stopped. The exception handling I added also stops the reader and that resulted in a double-stop.

To avoid that issue, I changed to handling bad_alloc's only as that was the exception I saw causing the original issue in practice.
John makes a good point above though. I can catch all exceptions and only stop the reader when it's not already been done.

VladLazar commented:
I implemented John's suggestion in the [last force push](https://github.com/redpanda-data/redpanda/compare/517783a703ede091ef50889e69238f7fc85bf962..a7576ba3d5d2abcf4586eaa362a4992844c3b21d). We catch all exceptions and if the reader still exists, we clean up.

jcsp reviewed (approved):
lgtm if it passes ci

VladLazar commented:
/backport v22.2.x

VladLazar commented:
/backport v22.1.x

VladLazar commented:
/backport v21.11.x
